### ISSUE TITLE
Allow some code to be silently pre-run in the `repl` app

### ISSUE THREAD

**Issue by jobovy** [jupyterlite/jupyterlite#508]
### Problem

Now that #498 has been merged and the `repl` app has been added, I'd like to raise the same issue as I raised in https://github.com/jtpio/replite/pull/24 to allow some code to be silently pre-run before handing over the app to the user. I'm opening a separate issue from #461, because I think the issues are slightly different between pre-running code in the `repl` and pre-running code in notebooks.

It seems to me that one of the main uses of the `repl` app is to embed it to allow people to try out code without having to install it. For example, @jtpio made a PR (https://github.com/numpy/numpy.org/pull/547) in numpy.org to use this `repl` app for `numpy`'s 'Try numpy' feature. However, because `pyodide` only contains a small subset of Python packages and, in particular, `pyodide` does not want to add Python packages that have a pure Python wheel, for most packages users have to do
```
import micropip # or using piplite
await micropip.install('PACKAGE')
```
before they can import PACKAGE. This code could be pre-populated using the `code` URL parameter, but I think that would be confusing for end users, because:
* Users expect this to be Python code, but this code can't actually be run in a Python interpreter;
* They are often novice Python users and will be very confused by the `await` keyword, which isn't part of Python

I'd love to use the `repl` app to make 'Try numpy' like REPLs for my Python packages' documentation, but showing the `import micropip...` code is a non-starter, since it will just lead to more confusion than it's worth among my users (who are often novice Python users). I think it's easy enough to add a disclaimer that the shown REPL isn't a _full_ Python environment, so not everything will work as normal.

### Proposed Solution

Add a URL parameter similar to the `prerun` parameter from https://github.com/jtpio/replite/pull/24 that allows code that is to be pre-run to be added at runtime.

#### Alternative solutions

* Don't allow arbitrary code to be pre-run, only install packages. E.g., my original PR in https://github.com/jtpio/replite/pull/24 added a `install` URL parameter that listed packages to be installed by `micropip`. @jtpio objected (reasonably) to this, because `micropip` is specific to the `pyolite` kernel, while the `repl` app supports many kernels. However, one could also make it a `python_install` URL parameter or `pyolite_install` parameter to make it clear that it's only to be used for `pyolite` kernels
* Somehow catch all instances of `import PACKAGE` that fail and try to do `import micropip; await micropip.install('PACKAGE'); install PACKAGE` as a fall-back. Downside of this is that it only allows packages to be installed that are on PyPI, not wheels hosted elsewhere.
* An extension that does the proposed solution. If this is the preferred option, I'd appreciate some pointers on how to do this, because I have no idea!

#### Reproducibility and user confusion

Some issues were raised in #461 that are worth addressing here. I think the main ones are:

1. Reproducibility: having code silently pre-run makes it confusing to the run same code on another instance of the `repl` that doesn't pre-run this code
2. You can only hide that you are running in `pyolite` rather than `Python` for so long and users might have to learn `pyolite` to import other packages that they might need (or fetch files over HTTPS etc.)

I think reproducibility is a lesser concern for the `repl` app, because there is no notebook to download and run elsewhere. And I think these `repl`s will generally be used in a context where it's clear to the user that it is a special case (e.g., the 'Try numpy' `repl`).

There's no way around 2. really, but I imagine that users will not generally be tempted to try to install packages that don't just load, say in the context of a `Try numpy`-like REPL. Embedded REPLs can just add a disclaimer about things not fully working as usual.

**Comment by bollwyvl** [comment]
If #503 lands, this would make a fine extension, but is again not something I feel we should enable by default in core... i don't even like that the _current_ implementation runs the code without explicitly asking:

```
https://jupyterlite.readthedocs.io/en/latest/_static/repl/index.html?kernel=javascript&code=alert(%27lol%20pwnt%27)
```

**Comment by psychemedia** [comment]
Minimising complexity and confusion for new users is key where the REPL, or demo notebooks, are used to onboard new users.

The `await` and `micropip` requirements distinguish the juptyerlite code from non-Jupyterlite Python environments; one way round this would be to support package installation via `%pip` as per https://github.com/jupyterlite/jupyterlite/issues/237 / This is magic after all, so the user needn't know what's going on underneath.

For pre-run code such as package imports, is there a way to do this verbosely, for example display a line prior to the first cursor saying something like *"This environment has been pre-configured by pre-importing the `X` and `Y` packages"*, or some formulation that also accommodates `from .... import ...` formulations.

**Comment by jtpio** [comment]
> For pre-run code such as package imports, is there a way to do this verbosely, for example display a line prior to the first cursor saying something like _"This environment has been pre-configured by pre-importing the `X` and `Y` packages"_, or some formulation that also accommodates `from .... import ...` formulations.

This is what the `?code=` query string parameter does at the moment in the `repl` app. For example it is used on https://numpy.org to automatically run `import numpy as np` on startup:

![image](https://user-images.githubusercontent.com/591645/154441305-0656a7d4-d305-4c85-a98c-a2cf7a82f0f1.png)

https://jupyterlite.readthedocs.io/en/latest/applications/repl.html#auto-execute-code-on-startup

**Comment by jobovy** [comment]
This isn't getting much traction it seems, but if anybody reading this is interested in this functionality, note that I've implemented it in [jobovy/jupyterlite:quiet-prerun](https://github.com/jobovy/jupyterlite/tree/quiet-prerun). This simply adds the `prerun` URL parameter to the `repl` app, as described in my proposed solution. It's a stopgap until somebody writes an extension to do this.

I also have an example of using this to deploy a REPL with a bunch of code to pre-run at [jobovy/galpy-repl](https://github.com/jobovy/galpy-repl), which builds the custom `jupyterlite` REPL and deploys it to AWS S3 together with a `repl` shortcut that includes the code to pre-run (see the README of that repository for more explanation of this).

**Comment by jtpio** [comment]
Thanks @jobovy for sharing this!

Do you think this could be implemented and distributed as a separate extension? Since there have been some concerns in this issue and related ones about reproducibility when hidden code is executed, maybe this could be experimented with first?

This also seems to be requested by several people, so having it easily installable via an extension will make it easier to reuse.

There is also an open PR to make the `repl` params extensible, which should make it easier to write such extension: https://github.com/jupyterlite/jupyterlite/pull/525

**Comment by martinRenou** [comment]
Note that this can be implemented using the IPython config `c.InteractiveShellApp.exec_lines`, see [IPython docs](https://ipython.readthedocs.io/en/stable/config/intro.html), so I wonder if we could allow passing a config to pyolite for IPython to use.

Although using the IPython config does not scale to other languages/kernels.

**Comment by jobovy** [comment]
> Do you think this could be implemented and distributed as a separate extension? Since there have been some concerns in this issue and related ones about reproducibility when hidden code is executed, maybe this could be experimented with first?

Yes, I think it would be great for this to be an extension. I'd be happy to work on this, but I have never written a `JupyterLab`-like extension before, so would appreciate some pointers (perhaps an example one that is similar, pointer to the relevant documentation).

> Note that this can be implemented using the IPython config `c.InteractiveShellApp.exec_lines`, see [IPython docs](https://ipython.readthedocs.io/en/stable/config/intro.html), so I wonder if we could allow passing a config to pyolite for IPython to use.
> 

This would be fine for my use case and, I imagine, a lot of other people's. It would be particularly useful to write the code to pre-run in a file rather than directly in the URL parameter, because the URL can otherwise become very complicated to write (same holds for the `code` URL parameter really). In my current hacky implementation of this, I have a file `prerun.py` that gets parsed and URL-encoded to my `prerun` URL parameter: see https://github.com/jobovy/galpy-repl and in particular https://github.com/jobovy/galpy-repl/blob/4e70582375abb47ea7598e891849358ff2e7831d/.github/workflows/deploy-repl-redirect.yml#L19. The advantage of using a file is that you can just run it as a regular Python/... file for testing.

**Comment by koenvo** [comment]
I hope I understand it correctly but when it's part of the config it would also don't require additional query parameters, rights? 

In my case I would like to create a Playground with all kinds of notebooks (pre-installed), but also have the packages that are required by those notebooks pre-installed. People will probably land on the Playground from all kind of places and I'm not sure how to make add the prerun query param - or with an additional redirect.

**Comment by jobovy** [comment]
Inspired by @jtpio's extension mentioned in https://github.com/jupyterlite/jupyterlite/issues/859#issuecomment-1500058694, I implemented a very simple extension with a `prerun` URL parameter that allows some code to be silently pre-run in the `repl` app:

https://github.com/jobovy/jupyterlite-repl-prerun

While it works okay, one issue with this extension is that one cannot combine the `prerun` and `code` URL parameters, because the `prerun` code runs after the `code` code (because the `jupyterlite-repl-prerun` extension runs after the `repl` extension). I guess #503 might solve that, although it would require adding dependencies between the URL parameters to make sure the `prerun` code is run before the `code` code (that is, one would have to be able to specify that `prerun`'s `onWidgetAdded` needs to run before `code`'s `onWidgetAdded`). To get around this for now, I've just added a `prerun-code` parameter to my extension that does the same as `code`.
